### ISSUE TITLE
Regarding Model Deployment on Test set

### ISSUE THREAD

**Issue by Juliyan96** [BioSystemsUM/DeepMol#68]
Dear Joao,
I have devloped my ML and DL models and I would like to know how to save my models with the approproate extensions and to use it on my external test set. I have attached my code for a ML and a DL model I have made.

Code ML model:
```
from boruta.boruta_py import BorutaPy
from deepmol.feature_selection.base_feature_selector import BorutaAlgorithm
from deepmol.compound_featurization import MorganFingerprint
from deepmol.compound_featurization import MACCSkeysFingerprint
from deepmol.loaders.loaders import CSVLoader
from deepmol.feature_selection import LowVarianceFS
from deepmol.splitters.splitters import SingletaskStratifiedSplitter
from deepmol.models.sklearn_models import SklearnModel
from deepmol.metrics.metrics import Metric
from deepmol.metrics.metrics_functions import precision_score, accuracy_score, confusion_matrix, classification_report, roc_auc_score
from deepmol.parameter_optimization.hyperparameter_optimization import HyperparameterOptimizerValidation
from deepmol.imbalanced_learn.imbalanced_learn import SMOTEENN
import numpy as np
from deepmol.standardizer.custom_standardizer import CustomStandardizer, heavy_standardisation
standardizer2 = CustomStandardizer(heavy_standardisation).standardize(dataset)
dataset = CSVLoader(dataset_path='/content/drive/MyDrive/2. Dataset_Deepmol.csv', mols_field='Smiles', labels_fields='Class', id_field='ID')
dataset = dataset.create_dataset()
dataset.get_shape()
dataset = MorganFingerprint(radius=2, size=1024).featurize(dataset)
print(dataset.X)
print(dataset.y)
print(dataset.mols)
print(dataset.get_shape())
dataset = LowVarianceFS(0.15).select_features(dataset)
dataset.get_shape()
splitter = SingletaskStratifiedSplitter()
train_dataset, valid_dataset, test_dataset = splitter.train_valid_test_split(dataset=dataset, frac_train=0.8,
                                                                             frac_valid=0.2, frac_test=0.0)
print('Train:')
train_dataset.get_shape()
print('\nValidation:')
valid_dataset.get_shape()
print('\nTest:')
test_dataset.get_shape()
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
model = SklearnModel(model=rf)
model.fit(train_dataset)
model.cross_validate(dataset, Metric(roc_auc_score), folds=10)
metrics = [Metric(roc_auc_score), Metric(precision_score), Metric(accuracy_score), Metric(confusion_matrix), 
           Metric(classification_report)]
print('Training Dataset: ')
train_score = model.evaluate(train_dataset, metrics)
print('Validation Dataset: ')
valid_score = model.evaluate(valid_dataset, metrics)
def rf_model_builder(n_estimators, max_features, class_weight, model_dir=None):
    rf_model = RandomForestClassifier(n_estimators=n_estimators, max_features=max_features, class_weight=class_weight)
    return rf_model#(rf_model, model_dir)
params_dict_rf = {"n_estimators": [10, 100],
                  "max_features": ["auto", "sqrt", "log2", None],
                  "class_weight": [{0: 1., 1: 1.}]}#, {0: 1., 1: 5}, {0: 1., 1: 10}]
                  #}
optimizer = HyperparameterOptimizerValidation (rf_model_builder)
best_rf, best_hyperparams, all_results = optimizer.hyperparameter_search(params_dict_rf, train_dataset, valid_dataset, Metric(roc_auc_score))
print(best_hyperparams)
print(best_rf)
best_rf.evaluate(valid_dataset, metrics)

```
DL Model:

```
from boruta.boruta_py import BorutaPy
from deepmol.feature_selection.base_feature_selector import BorutaAlgorithm
from deepmol.loaders.loaders import CSVLoader
from deepmol.feature_selection import LowVarianceFS
from deepmol.splitters.splitters import SingletaskStratifiedSplitter
from deepmol.metrics.metrics import Metric
from deepmol.metrics.metrics_functions import roc_auc_score, precision_score, accuracy_score, confusion_matrix, classification_report
from deepmol.imbalanced_learn.imbalanced_learn import SMOTEENN
from deepmol.standardizer.custom_standardizer import CustomStandardizer, heavy_standardisation
import numpy as np

standardizer2 = CustomStandardizer(heavy_standardisation).standardize(dataset)
dataset = CSVLoader(dataset_path='/content/drive/MyDrive/2. Dataset_Deepmol.csv', mols_field='Smiles', labels_fields='Class', id_field='ID')
dataset = dataset.create_dataset()
dataset.get_shape()
from deepchem.models import GraphConvModel
from deepmol.models.deepchem_models import DeepChemModel
from deepmol.splitters.splitters import SingletaskStratifiedSplitter
from deepmol.compound_featurization.deepchem_featurizers import ConvMolFeat
#from deepmol.compound_featurization import ConvMolFeat

ds = ConvMolFeat().featurize(dataset)
splitter = SingletaskStratifiedSplitter()
train_dataset, valid_dataset, test_dataset = splitter.train_valid_test_split(dataset=ds, frac_train=0.7, frac_valid=0.3,
                                                                             frac_test=0)
print('Train:')
train_dataset.get_shape()
print('\nValidation:')
valid_dataset.get_shape()
graph = GraphConvModel(n_tasks=1, mode='classification')
model_graph = DeepChemModel(graph)
model_graph.fit(train_dataset)
valid_preds = model_graph.predict(valid_dataset)
metrics = [Metric(roc_auc_score), Metric(precision_score), Metric(accuracy_score), Metric(confusion_matrix), 
           Metric(classification_report)]
print('Training Dataset: ')
train_score = model_graph.evaluate(train_dataset, metrics)
print('Validation Dataset: ')
valid_score = model_graph.evaluate(valid_dataset, metrics)
```
Thank you!

**Comment by jcapels** [comment]
Hi, there, sorry for the late reply.

We didn't solve this issue, so I think you can do the following for scikit-learn models:

```python
import pickle

filename = 'best_rf.pkl'
with open(filename, 'wb') as file:
    pickle.dump(best_rf, file)
```

and load like this:
```python
import pickle

filename = 'best_rf.pkl'
with open(filename, 'rb') as file:
    best_rf = pickle.load(file)
```

Whereas for the deep learning model:

```python
folder_name = 'best_graph_conv'
best_model.model.save_checkpoint(model_dir=folder_name)

graph = GraphConvModel(n_tasks=1, mode='classification')
loaded_model = DeepChemModel(graph, model_dir=None, epochs=5)
loaded_model.model.get_checkpoints(model_dir=folder_name)
loaded_model.predict(test_dataset)
```
